[GPII-3341]: Mutual TLS for CouchDB

natarajaya opened the pull request:
Currently mutual TLS is set to `STRICT` mode for entire cluster, but was disabled for CouchDB with DestinationRule.

Unfortunately, when turned on – it makes CouchDB split brain problem (GPII-3624) to occur almost on every run.
The reason is that `couchdb-statefulset-assembler` puts too much trust in cluster DNS (details in PR https://github.com/gpii-ops/couchdb-statefulset-assembler/pull/1).

I'll work on deployment plan for this after `couchdb-statefulset-assembler` issues are resolved.

stepanstipl commented:
Istio can't be simply enabled and `DestinationRule` rule removed for CouchDB - this would cause CouchDB pods to fail to see each other during the rolling restart - ie. pods on Istio wouldn't be able to talk to non-istio pods.

> Unfortunately, when turned on – it makes CouchDB split brain problem (GPII-3624) to occur almost on every run.

Are you referring here to creating a new cluster or upgrading an existing cluster?

Ports of CouchDB services should be renamed to match istio naming convention `<protocol>[-<suffix>]` - https://istio.io/docs/setup/kubernetes/additional-setup/requirements/.

Also Istio will probably struggle with the headless service that CouchDB chart uses, quick search suggests it might need an additional serviceentry - https://github.com/istio/istio/issues/7495.

At last -  adding a `COUCHDB_CLUSTER_SIZE` to spec might work as a short term solution, but otherwise would cause a whole cluster restart on operations like scaling up, so probably quite undesirable.

natarajaya commented:
> Istio can't be simply enabled and DestinationRule rule removed for CouchDB - this would cause CouchDB pods to fail to see each other during the rolling restart

@stepanstipl As stated in PR description, I have not worked on deployment plan yet, so I don't know if this is the case or not, but thanks for your notice! It works fine when cluster is being created from scratch, so worst case scenario – we can announce short downtime to fully redeploy CouchDB modules.

> Are you referring here to creating a new cluster or upgrading an existing cluster?

Both of scenarios can be affected.

> Ports of CouchDB services should be renamed to match istio naming convention 

This guide does not exist in [docs for 1.0](https://archive.istio.io/v1.0/docs/setup/kubernetes/). I am fine renaming them, but not sure if it makes any difference while we are running GKE 1.12 with Istio 1.0.6.

> Also Istio will probably struggle with the headless service that CouchDB chart uses

Everything works just fine with existing services (at least smoke tests are passing and I have not noticed any unexpected behaviors). Also issue that you mentioned is for Istio 0.8.

> At last - adding a COUCHDB_CLUSTER_SIZE to spec might work as a short term solution, but otherwise would cause a whole cluster restart on operations like scaling up, so probably quite undesirable

I think the proper way of dealing with this is to populate env vars from secret, so when scaling up only secret is updated and stateful set remains unchanged. I tested this behavior and looks like it works as expected – no unnecessary pod restarts.

natarajaya commented:
> Istio can't be simply enabled and DestinationRule rule removed for CouchDB - this would cause CouchDB pods to fail to see each other during the rolling restart

@stepanstipl Update: I just run code from this PR against existing cluster that was earlier deployed from master – it went without any issue. Also, I don't see any failed health checks for flowmanager / preferences in the logs.

`tls-check` looks good:

```bash-4.4# istioctl authn tls-check couchdb-svc-couchdb.gpii.svc.cluster.local

HOST:PORT                                           STATUS     SERVER     CLIENT     AUTHN POLICY     DESTINATION RULE
couchdb-svc-couchdb.gpii.svc.cluster.local:5984     OK         mTLS       mTLS       default/         default/istio-system```

So I expect if any – minimal downtime. I'll repeat this exercise again tomorrow, but running smoke tests in background to get better numbers.

stepanstipl commented:
Thanks @natarajaya 

> so worst case scenario – we can announce short downtime to fully redeploy CouchDB modules.

This can be avoided by doing multiple-step release:
1. Keep `DestinationRule` in place
1. Create `Policy` to relax the mTLS requirement using `mtls: mode: PERMISSIVE` - this would allow pods with Istio to receive traffic from pods without
1. Enable Istio on CouchDB
1. Remove `DestinationRule` & remove the `Policy`

> @stepanstipl Update: I just run code from this PR against existing cluster that was earlier deployed from master – it went without any issue. Also, I don't see any failed health checks for flowmanager / preferences in the logs.

This is pretty interesting since definitely non-istio pods would not be able to talk to istio pods during the rolling update. This would also be true for our services already running on istio. I would expect the rolling update to get stuck, but I guess if it went through it means that our health-check does not reflect really whether the CouchDB node can talk to other nodes and/or is caught-up with replication. _I checked: it doesn't, the `_up` endpoint happily returns `"status":"ok"` even if it can't see any other nodes of the cluster._

> This guide does not exist in docs for 1.0. I am fine renaming them, but not sure if it makes any difference while we are running GKE 1.12 with Istio 1.0.6.

It's here in 1.0 docs: https://archive.istio.io/v1.0/docs/setup/kubernetes/spec-requirements/ It affects how Istio does routing and it's important for metrics reporting.

> Everything works just fine with existing services (at least smoke tests are passing, logs look good and I have not noticed any unexpected behaviors). Also, issue that you mentioned is for Istio 0.8.

That sounds good. Can we somehow verify that all the intra-cluster replication is working as expected?

Relying only on the `_up`, and incl. running our tests, we could end up with 3 nodes of CouchDB which work, but can't talk to each other and we wouldn't know about it. (And I believe that's when the direct pod-to-pod communication is used - for node replication).

One basic way is to query the `/_membership` and verify all the nodes are listed under `"all_nodes". I've also tried to use the `/{db}/_sync_shards` endpoint but that doesn't seem to work for the purpose (ie. will return `{"ok":true}` even if the nodes can't talk to each other). It would be nice to have some better way to verify that all nodes are in-sync, like based on some revision (I don't know how to do that).

I've tried to do a little test and **blocked all the traffic between individual CouchDB pods**. The pods were still reporting as healthy and **both the smoke tests passed**.

I think this is a significant problem with our monitoring, out of the scope of this ticket, but something we should definitely look at. **Currently we could easily be running DB that does not replicate correctly between nodes and we would not know about it.**

In any case, we should make sure that the replication traffic between CouchDB nodes works fine after this change.

> I think the proper way of dealing with this is to populate env vars from secret

I like this idea, it's definitely an improvement 👍 good thinking ;). I would move this to `ConfigMap` as the cluster size does not need to be secret (and it would avoid the necessity of base64 encoding) and also should not depend on the `{{- if .Values.createAdminSecret -}}` condition.

natarajaya commented:
I moved all changes related to CouchDB split brain issues to a separate PR (#456).

@stepanstipl Thanks for your research! Seems like migration with intermediate `PERMISSIVE ` mode is the seamless way to go. However, I propose to not overcomplicate things and deploy this together with #456, since it already needs a planned downtime.

> It's here in 1.0 docs It affects how Istio does routing and it's important for metrics reporting

Cool, thanks! Discrepancy in Istio docs for different versions is a little bit disturbing.

> Currently we could easily be running DB that does not replicate correctly between nodes and we would not know about it

This sounds pretty bad. I added action item to discuss this further.

> I would move this to ConfigMap as the cluster size does not need to be secret

Good suggestion! ConfigMap is definitely a better place for this to live.

> In any case, we should make sure that the replication traffic between CouchDB nodes works fine after this change.

Checking CouchDB internal replication status is somewhat non-trivial task. But I verified, and at-least node-to-node connectivity is not affected by this migration.

natarajaya commented:
> In any case, we should make sure that the replication traffic between CouchDB nodes works fine after this change.

@stepanstipl Adding `http-` prefix to 5984 port breaks node-to-node connectivity. When I prefix it as `tcp-` everything goes back to normal. But docs say: "If the port name does not begin with a recognized prefix or if the port is unnamed, traffic on the port is treated as plain TCP traffic ". I'll try to look some more into `http-` prefix issue, but otherwise there is no need to change ports naming.

stepanstipl commented:
Thanks @natarajaya 

> I moved all changes related to CouchDB split brain issues to a separate PR (#456).

👍 

> However, I propose to not overcomplicate things and deploy this together with #456, since it already needs a planned downtime.

Yeah, I think it's ok, as long as we're aware & acknowledge the short downtime.

> I'll try to look some more into http- prefix issue, but otherwise there is no need to change ports naming.

👍, I think it's worth looking into. This changes how Istio intercepts and routes traffic (there are some additional limits and lot of features not available for tcp) and also what metrics you have available - if the port is not correctly named with `http-`, you won't see http L7 metrics - like response codes

natarajaya commented:
> I think it's worth looking into. This changes how Istio intercepts and routes traffic 

@stepanstipl HTTP mode is now working as expected after I formatted ServiceEntries you suggested correctly!

I am still struggling to find a workaround for headless service issue that we discussed: https://github.com/istio/istio/issues/12551

natarajaya commented:
Ok, I think this is finally ready to be deployed. I'll post deployment plan into the ticket.

It turned out that main reason for clustering issues was Erlang's inability to talk to other nodes using its TCP ports. This is solved once I exposed ports on headless service and added them to ServiceEntry.

Lack of pod-to-self communication still bothers me a little, but I double (triple) checked and CouchDB seems to be fine with that. There is no official solution for that issue and the only thing that worked looks pretty ugly:

```
- name: istio-iptables-hack
  env:
    - name: POD_IP
      valueFrom:
        fieldRef:
          fieldPath: status.podIP
  image: gcr.io/gke-release/istio/proxy_init:1.0.6-gke.3
  securityContext:
    capabilities:
      add:
      - NET_ADMIN
  command:
    - sh
    - -c
    - |
      iptables -t nat -I ISTIO_OUTPUT -d ${POD_IP} -j RETURN &&
      while true; do sleep 10; done
```

Adding this to stateful set containers drops pod IP from `ISTIO_OUTPUT` iptables chain and allows pod to communicate with itself directly. Which, in theory, should still be secure, since pod IP is assigned to network interface that is attached to container (pod-to-self traffic is local). I used [proxy_init](https://github.com/istio/istio/blob/master/pilot/docker/Dockerfile.proxy_init) as smallest image with iptables binary which is already present on the host.

natarajaya commented:
@stepanstipl I tried solution from your gist and it did not work – I am not sure how applying service-specific policy to clusterIP service can affect pods communication under headless service DNS (also I don't understand your comment about port 15000). I tried some other options to enable permissive mTLS mode as well, as long as Istio's auto-injection is enabled for stateful set, pod-to-self connectivity is affected.

mrtyler commented on shared/charts/couchdb-prometheus-exporter/templates/deployment.yaml:
```diff
@@ -10,7 +10,7 @@ spec:
         app: couchdb-prometheus-exporter
       annotations:
-        sidecar.istio.io/inject: "false"
+        traffic.sidecar.istio.io/excludeOutboundIPRanges: "169.254.169.254/32"
```
Might be worth mentioning what this is for?

If so, please update the same line in the flowmanager and preferences charts.

natarajaya replied:
This is to allow metadata traffic. Once we enable sidecar auto-injection for couchdb-prometheus-exporter, all destinations that are not explicitly allowed with ServiceEntries, are forbidden. And Istio 1.0.6 is still affected by FQDN issue (https://github.com/istio/istio/issues/14404), so allowing metadata traffic is impossible via ServiceEntry, hence this hack.

mrtyler replied:
Great, thanks. Please add that explanation as a comment to all three charts.

stepanstipl commented:
> @stepanstipl I tried solution from your gist and it did not work – I am not sure how applying service-specific policy to clusterIP service can affect pods communication under headless service DNS

Istio authentication policies apply to requests that a service receives, in the context of the policy from the gist, the target selector:
```yaml
  targets:
  - name: couchdb-svc-couchdb
```
will apply the policy to all pods that are members of the `couchdb-svc-couchdb` service. This is required to relax the `mTLS` policy to `PERMISSIVE` on the CouchDB pods.

> (also I don't understand your comment about port 15000).

Should have said `15001`.

natarajaya commented:
> will apply the policy to all pods that are members of the couchdb-svc-couchdb service. This is required to relax the mTLS policy to PERMISSIVE on the CouchDB pods.

@stepanstipl This does not seem to be the case for pod-to-pod communication.

> Should have said 15001

What do you mean?

stepanstipl commented:
>> will apply the policy to all pods that are members of the couchdb-svc-couchdb service. This is required to relax the mTLS policy to PERMISSIVE on the CouchDB pods.

> @stepanstipl This does not seem to be the case for pod-to-pod communication.

@natarajaya It's not easy to convince you :) -> https://github.com/stepanstipl/test-istio-pod-2-pod

>> Should have said 15001

> What do you mean?

Envoy proxy is accepting all traffic on port `15001`, so in the context of relaxing mTLS to permissive mode this means that any pod will be able to connect to the service, even pod outside of the Istio mesh, by using port `15001`.

natarajaya commented:
> It's not easy to convince you :) -> https://github.com/stepanstipl/test-istio-pod-2-pod

@stepanstipl Ok, you convinced me after all :D But I swear I tried exactly the same ServiceEntry and Policy configuration before, and it did not affect pod-to-pod communication in any way. Not sure what was missing (maybe another Istio inconsistency simlar to what we observe with TLS routing?), but if I add the following code to the changes to CouchDB chart from this PR:
```
apiVersion: authentication.istio.io/v1alpha1
kind: Policy
metadata:
  name: {{ template "couchdb.fullname" . }}
spec:
  peers:
  - mtls:
      mode: PERMISSIVE
  targets:
  - name: {{ template "couchdb.svcname" . }}
```
I can confirm that tls-check reports permissive configuration for pods DNS and pod-to-self connectivity is restored! So looks like we have less ugly workaround for pod-to-self communication in case it is needed, good job!
